laravel/valet
# Release Version 4

**Done:**
- Re-work basic Valet driver: #1311 
- Add a hook before `frontControllerPath` so drivers don't do so much in there (#1319)
- Extract server.php to mainly a class (#1319), and test it (https://github.com/laravel/valet/pull/1318/commits/65b629db7fa559448baf9eb0b7a1d061e2398e14)
- Add type hints and return types (#1321)
- Drop all the param/return notes in the docblocks (like the rest of Laravel is: https://twitter.com/taylorotwell/status/1610338750843740160)
- Convert to constructor promotion (https://github.com/laravel/valet/pull/1318/commits/75d43a42ea0396c2346f5bbee11d23c5ebfdbf6b)
- Test `Server.php` (the class)  (https://github.com/laravel/valet/pull/1318/commits/65b629db7fa559448baf9eb0b7a1d061e2398e14)
- Make an automated upgrade step that updates already-published files like `SampleValetDriver` to have the right method signature, etc. (https://github.com/laravel/valet/pull/1318/commits/9b09621e5cb00e9a2d7b27df7179be06531d6027)
- Drop extensions
- Test every single feature/command since all these changes; anything that's broken, write a test before fixing
- Make a `valet status` command (https://github.com/laravel/valet/pull/1329, https://github.com/laravel/valet/issues/251)
- Upgrade to ngrok 3.1 (x86, see #1324)
- Test CLI commands (`CliTest`)
- #1340 (including updating ngrok to use Brew, allowing for both arm and x64 versions)
- Give drivers access to check if a given Composer package is installed (like how Sculpin does it) and potentially simplify some drivers to just use that (addresses #1342, solved in #1345)
- If global PHP is less than 8.0, update `valet` (bash file) to look for a brew-installed-but-not-linked PHP executable and use that throughout the rest of the `valet` file (https://github.com/laravel/valet/pull/1346)
- Check if site isolation can work on 7.0-7.3 given current `Server.php` code (update: It works on 7.1-7.4, dropping 7.0 support because I'm not gonna kill all our return types)
- Add `.valetrc`, update calls to `.valetphprc` to also support the `php` key of `.valetrc` (#1347)
- Better manage running services to make sure they're running as the right user (see @nicoverbruggen notes here about how he's doing that in PHPMon: https://github.com/laravel/valet/pull/1329#issuecomment-1366102502) : https://github.com/laravel/valet/pull/1348
- Finish making Expose work (https://github.com/laravel/valet/pull/1349)
- If someone edited their `SampleValetDriver`, back up the old one before putting in a new one (https://github.com/laravel/valet/pull/1318/commits/15cf3fefaf540fc772339d9c9f8040fa902e9ae8)
- Un-secure when un-linking (https://github.com/laravel/valet/pull/1364)

----

**To do after v4 release:**
- Update `valet status` to check to make sure the certificate authority isn't expired
- Look into @nicoverbruggen's trust CLI app: https://github.com/laravel/valet/issues/1226
- Build a system where the Mac-specific parts of the codebase are separate, so the Linux and Windows ports can extend this package (#1323)
- Add other powers to `.valetrc`
    - E.g. specify nginx includes to use when generating the Nginx config for this site
        - Would require a command that does `valet apply-config` or something, that e.g. re-generates nginx config files... 
        - Consider: is it easier to have a potential custom stub directory for each site, either in the site or globally? Maybe allows the stubs/.valetrc type modifications to be in source control that waybasically `install` just for this one site
- Default home page Marcel & Matt made
- Consider: Plugin system suggested by Marcel (https://twitter.com/marcelpociot/status/1604603384203452416)

**To document:**
- `.valetrc` and deprecated `.valetphprc`
- `set-ngrok-token` command
- New `valet share` that's ngrok + expose, and the `valet share-tool` command
- Custom stubs https://github.com/laravel/valet/pull/1238/files
- upgrade process to v4
- todo any other features/tweaks made here
